[Translation] Document ICU MessageFormat (#11523)
symfony/symfony-docs: translation/message_format.rst

.. index::
    single: Translation; Message Format

> How to translate messages using the ICU MessageFormat

Reviewer: This should use the "Title Case".

=====================================================

.. versionadded:: 4.2

   Support for ICU MessageFormat was introduced in Symfony 4.2.

Messages (i.e. strings) in applications are almost never completely static.

> They contain variables or other complexer logic like pluralization. In order to

Reviewer: `other complexer logic` -> `other complex logic` sounds better to me

handle this, the Translator component supports the `ICU MessageFormat`_ syntax.

Using the ICU Message Format
----------------------------

In order to use the ICU Message Format, the :ref:`message domain
<using-message-domains>` has to be suffixed with ``intl-icu``:

======================  ===============================
Normal file name        ICU Message Format filename
======================  ===============================
``messages.en.yml``     ``messages+intl-icu.en.yml``
``messages.fr_FR.xlf``  ``messages+intl-icu.fr_FR.xlf``
``admin.en.yml``        ``admin+intl-icu.en.yml``
======================  ===============================

All messages in this file will now be processed by the
:phpclass:`MessageFormatter` during translation.

.. _component-translation-placeholders:

Message Placeholders
--------------------

> The basic usage of the MessageFormat allows you to use placeholder (called

Reviewer: `placeholder` -> `placeholders`

*arguments* in ICU MessageFormat) in your messages:

.. configuration-block::

    .. code-block:: yaml

        # translations/messages.en.yaml
        say_hello: 'Hello {name}!'

    .. code-block:: xml

        <!-- translations/messages.en.xlf -->
        <?xml version="1.0"?>
        <xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2">
            <file source-language="en" datatype="plaintext" original="file.ext">
                <body>
                    <trans-unit id="say_hello">
                        <source>say_hello</source>
                        <target>Hello {name}!</target>
                    </trans-unit>
                </body>
            </file>
        </xliff>

    .. code-block:: php

        // translations/messages.en.php
        return [
            'say_hello' => "Hello {name}!",
        ];

Everything within the curly braces (``{...}``) is processed by the formatter

> and replaced by it's placeholder::

Reviewer: `it's` -> `its`

>     // ...

Reviewer 2: ```suggestion

```
not needed IMHO

>     echo $translator->trans('say_hello', ['name' => 'Fabien']); // Hello Fabien!

Reviewer: Let's align the comments at the end of these two lines to improve readability a bit.

>     echo $translator->trans('say_hello', ['name' => 'Fabien']); // Hello Fabien!

Reviewer 2: In almost all following example we put the result first:
```php
// I am the result
$i_am_the_php_code
```

so I propose to put the result above the line of php code

    echo $translator->trans('say_hello', ['name' => 'Symfony']); // Hello Symfony!

Selecting Different Messages Based on a Condition
-------------------------------------------------

The curly brace syntax allows to "modify" the output of the variable. One of

> these functions is the ``switch`` function. It acts like PHP's `switch statement`_

Reviewer 2: based on the following examples, yes 👍

and allows to use different strings based on the value of the variable. A
typical usage of this is gender:

.. configuration-block::

    .. code-block:: yaml

        # translations/messages.en.yaml
        invitation_title: >
            {organizer_gender, select,
                female {{organizer_name} has invited you for her party!}
                male   {{organizer_name} has invited you for his party!}
                other  {{organizer_name} have invited you for their party!}
            }

    .. code-block:: xml

        <!-- translations/messages.en.xlf -->
        <?xml version="1.0"?>
        <xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2">
            <file source-language="en" datatype="plaintext" original="file.ext">
                <body>
                    <trans-unit id="invitation_title">
                        <source>invitation_title</source>

>                         <target>{organizer_gender, select, female {{organizer_name} has invited you for her party!} male   {{organizer_name} has invited you for his party!} other {{organizer_name} have invited you for their party!}}</target>

Reviewer 2: Maybe its just a rendering problem, but it looks like there are many whitespaces after `male` and not after `other`

                    </trans-unit>
                </body>
            </file>
        </xliff>

    .. code-block:: php

        // translations/messages.en.php
        return [
            'invitation_title' => '{organizer_gender, select,
                female {{organizer_name} has invited you for her party!}
                male   {{organizer_name} has invited you for his party!}
                other  {{organizer_name} have invited you for their party!}
            }',
        ];

This might look very complex. The basic syntax for all functions is
``{variable_name, function_name, function_statement}``. In this case, the
function name is ``select`` and its statement contains the "cases" of this
select. This function is applied over the ``organizer_gender`` variable::

>     // ...

Reviewer 2: ```suggestion

```
could be removed IMHO

    echo $translator->trans('invition_title', [
        'organizer_name' => 'Ryan',
        'organizer_gender' => 'male',
    ]);

    // prints "John & Jane have invited you for their party!"
    echo $translator->trans('invition_title', [
        'organizer_name' => 'John & Jane',
        'organizer_gender' => 'not_applicable',
    ]);

The ``{...}`` syntax alternates between "literal" and "code" mode. This allows
you to use literal text in the select statements:

#. The first ``{organizer_gender, select, ...}`` block starts the "code" mode,
   which means ``organizer_gender`` is processed as a variable.
#. The inner ``{... has invited you for her party!}`` block brings you back in
   "literal" mode, meaning the text is not processed.
#. Inside this block, ``{organizer_name}`` starts "code" mode again, allowing
   ``organizer_name`` to be processed as variable.

.. tip::

    While it might seem more logical to only put ``her``, ``his`` or ``their``
    in the switch statement, it is better to use "complex arguments" at the
    outermost structure of the message. The strings are in this way better
    readable for translators and, as you can see in the ``other`` case, other

>     parts of the sentence mgith be influenced by the variable.

Reviewer: `mgith` -> `might`

>     parts of the sentence mgith be influenced by the variable.

Reviewer 2: ```suggestion
    parts of the sentence might be influenced by the variables.
```

Pluralization
-------------

Another interesting function is ``plural``. It allows you to
handle pluralization in your messages (e.g. ``There are 3 apples`` vs

> ``There is one apple``). The function looks very similair to the ``select`` function:

Reviewer: `similair` -> `similar`

.. configuration-block::

    .. code-block:: yaml

        # translations/messages.en.yaml

>         nr_of_apples: >

Reviewer: `nr_of_apples` -> `num_of_apples`

            {apples, plural,

>                 =0    {There are no apples :(}

Reviewer: Let's remove the `:(` at the end of this phrase because it's confusing with the ICU syntax braces, etc.

                one   {There is one apple...}
                other {There are # apples!}
            }

    .. code-block:: xml

        <!-- translations/messages.en.xlf -->
        <?xml version="1.0"?>
        <xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2">
            <file source-language="en" datatype="plaintext" original="file.ext">
                <body>
                    <trans-unit id="nr_of_apples">
                        <source>nr_of_apples</source>
                        <target>{apples, plural, =0 {There are no apples :(} one {There is one apple...} other {There are # apples!}}</target>
                    </trans-unit>
                </body>
            </file>
        </xliff>

    .. code-block:: php

        // translations/messages.en.php
        return [
            'nr_of_apples' => '{apples, plural,
                =0    {There are no apples :(}
                one   {There is one apple...}
                other {There are # apples!}
            }',
        ];

Pluralization rules are actually quite complex and differ for each language.
For instance, here is the mathematical representation of the Russian
pluralization rules::

>     (($number % 10 == 1) && ($number % 100 != 11))

Reviewer 2: TBH this looks very complex and I can't get the value of having this example in the docs and don't understand it anyway 😄 WDYT?

Author: After second thought, I  agree. I've now explained it shortly in text. The main reason why I want to include this (and why it was previously included), is that pluralization for english speaking people (and most other countries) seems relatively simple: one (apple) and more than one (apples). So a simple `{number, select, =0 {...}, =1 {...}, other {...}}` seems to be enough. But the reality is that pluralization in many languages is actually very complex and cannot, in any reasonable way, be done using the `select` function.

        ? 0
        : ((($number % 10 >= 2)
            && ($number % 10 <= 4)
            && (($number % 100 < 10)
            || ($number % 100 >= 20)))
                ? 1
                : 2
    );

In order to facilitate this, the possible cases in the ``plural`` function are
also different for each language. For instance, Russian has ``one``, ``few``,
``many`` and ``other``, while English has only ``one`` and ``other``. The full
list of possible cases can be found in Unicode's `Language Plural Rules`_
document. By prefixing with ``=``, you can match exact values (like ``0`` in
the above example).

Usage of this string is the same as with variables and select::

>     // ...

Reviewer 2: ```suggestion

```

same as above

    ``select`` as outermost function:

    .. code-block:: text

		{gender_of_host, select, 
            female {
                {num_guests, plural, offset:1 
                =0    {{host} does not give a party.}
                =1    {{host} invites {guest} to her party.}
                =2    {{host} invites {guest} and one other person to her party.}
                other {{host} invites {guest} and # other people to her party.}}
            }
            male {
                {num_guests, plural, offset:1 
                =0    {{host} does not give a party.}
                =1    {{host} invites {guest} to his party.}
                =2    {{host} invites {guest} and one other person to his party.}
                other {{host} invites {guest} and # other people to his party.}}
            }
            other {
                {num_guests, plural, offset:1 
                =0    {{host} does not give a party.}
                =1    {{host} invites {guest} to their party.}
                =2    {{host} invites {guest} and one other person to their party.}
                other {{host} invites {guest} and # other people to their party.}}
            }
        }

> Other Placeholder Functions

Reviewer 2: ```suggestion
Additional Placeholder Functions
```

??

---------------------------

> Besides these, the MessageFormat comes with a couple other interesting functions.

Reviewer 2: ```suggestion
Besides these, the ICU MessageFormat comes with a couple other interesting functions.
```

??

Ordinal
~~~~~~~

> Similair to ``plural``, ``selectordinal`` allows you to use numbers as ordinal scale:

Reviewer: `Similair` -> `Similar`

.. configuration-block::

    .. code-block:: yaml

        # translations/messages.en.yaml
        finish_place: >
            You finished {place, selectordinal,
                one {#st}
                two {#nd}
                few {#rd}
                other {#th}

>             }!

Reviewer 2: Can you explain why to use the `!` here at the end? Or did I miss something? 🤔

Author: Part of the message ("You finished first!"). If you think it's confusing I'll remove it :)

Reviewer 2: Oh right, na let’s keep it 👍🏻

> 

Reviewer: I agree. Using "1, 2, 3" makes it easier to understand than using "one, two, few"  ("few" specially is so vague that it's confusing).

> 

Reviewer 3: Well, actually they're not the same and wasn't my point :P

My suggestion is to split the example; the first one can be simplified to just using 'ordinal' and let ICU handle all that formatting. A better example for selectordinal is to use something like my example. Although the other-case should perhaps be adjusted to something like this:
`other {didn't finish in the top three, but on # place}` (haven't tested that this works though). It could otherwise also be implemented using 'count'.

Btw not just 'few' can be confusing, the one and two may also be confusing. Some languages apply the one and two rule to any number ending in 1 and 2 (i.e. 22 => twenty-second), for either or both of 'count' and 'ordinal'.

> 

Author: Yeah, keywords might be confusing (but also not that much imo). But that's how the standard works, advising a bad practice to use `=...` is not a good idea imo.

I wasn't aware of the `ordinal` function. As I want to show how this `selectordinal` works, I've now extended this example by also adding a line showing the "simpler" `ordinal` case used for the same message. I can't think of a good example showing "one", "two", "few" in `selectordinal` without it being simplified to `ordinal`.

    .. code-block:: xml

        <!-- translations/messages.en.xlf -->
        <?xml version="1.0"?>
        <xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2">
            <file source-language="en" datatype="plaintext" original="file.ext">
                <body>
                    <trans-unit id="finish_place">
                        <source>finish_place</source>
                        <target>{apples, plural, =0 {There are no apples :(} one {There is one apple...} other {There are # apples!}}</target>
                    </trans-unit>
                </body>
            </file>
        </xliff>

    .. code-block:: php

        // translations/messages.en.php
        return [
            'finish_place' => 'You finished {place, selectordinal,
                one {#st}
                two {#nd}
                few {#rd}
                other {#th}
            }!',
        ];

.. code-block:: php

>     // ...

Reviewer 2: ```suggestion

```

The possible cases for this are also shown in Unicode's `Language Plural Rules`_ document.

Date and Time
~~~~~~~~~~~~~

The date and time function allows you to format dates in the target locale
using the :phpclass:`IntlDateFormatter`:

.. configuration-block::

    .. code-block:: yaml

        # translations/messages.en.yaml
        published_at: 'Published at {time, date} - {time, time, short}'

    .. code-block:: xml

        <!-- translations/messages.en.xlf -->
        <?xml version="1.0"?>
        <xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2">
            <file source-language="en" datatype="plaintext" original="file.ext">
                <body>
                    <trans-unit id="published_at">
                        <source>published_at</source>
                        <target>Published at {time, date} - {time, time, short}</target>
                    </trans-unit>
                </body>
            </file>
        </xliff>

    .. code-block:: php

        // translations/messages.en.php
        return [
            'published_at' => 'Published at {time, date} - {time, time, short}',
        ];

.. code-block:: php

>     // ...

Reviewer 2: ```suggestion

```

    // prints "Published at Jan 25, 2019 - 11:30 AM"

>     echo $translator->trans('published_at', ['time' => new \DateTime('2019-01-25 11:30:00')]);

Reviewer: Using `time` as the variable name is very confusing later when using it in the translation; example: `{time, time, short}` <- which is the variable and which is the function name? We could use a long variable name instead: `publication_date`

.. index::
    single: Translation; Message Format

How to Translate Messages using the ICU MessageFormat
=====================================================

.. versionadded:: 4.2

   Support for ICU MessageFormat was introduced in Symfony 4.2.

Messages (i.e. strings) in applications are almost never completely static.
They contain variables or other complex logic like pluralization. In order to
handle this, the Translator component supports the `ICU MessageFormat`_ syntax.

.. tip::

    You can test out examples of the ICU MessageFormatter in this `online editor`_.

Using the ICU Message Format
----------------------------

In order to use the ICU Message Format, the :ref:`message domain

> <using-message-domains>` has to be suffixed with ``intl-icu``:

Reviewer 4: Missing the + no?

+intl-icu

======================  ===============================
Normal file name        ICU Message Format filename
======================  ===============================
``messages.en.yml``     ``messages+intl-icu.en.yml``
``messages.fr_FR.xlf``  ``messages+intl-icu.fr_FR.xlf``
``admin.en.yml``        ``admin+intl-icu.en.yml``
======================  ===============================

All messages in this file will now be processed by the
:phpclass:`MessageFormatter` during translation.

.. _component-translation-placeholders:

Message Placeholders
--------------------

The basic usage of the MessageFormat allows you to use placeholders (called
*arguments* in ICU MessageFormat) in your messages:

.. configuration-block::

    .. code-block:: yaml

>         # translations/messages.en.yaml

Reviewer 4: Shouldn’t the file be name with the intl icu suffix?
Or i miss something?

using the :phpclass:`IntlDateFormatter`:

.. configuration-block::

    .. code-block:: yaml

        # translations/messages.en.yaml
        published_at: 'Published at {publication_date, date} - {publication_date, time, short}'

    .. code-block:: xml

        <!-- translations/messages.en.xlf -->
        <?xml version="1.0"?>
        <xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2">
            <file source-language="en" datatype="plaintext" original="file.ext">
                <body>
                    <trans-unit id="published_at">
                        <source>published_at</source>
                        <target>Published at {publication_date, date} - {publication_date, time, short}</target>
                    </trans-unit>
                </body>
            </file>
        </xliff>

    .. code-block:: php

        // translations/messages.en.php
        return [
            'published_at' => 'Published at {publication_date, date} - {publication_date, time, short}',
        ];

The "function statement" for the ``time`` and ``date`` functions can be one of
short, medium, long or full, as documented on PHP.net.

.. code-block:: php

>     // prints "Published at Jan 25, 2019 - 11:30 AM"

Reviewer 4: Maybe puts prints next to or after the echo?

Author: Hmm, I don't have a good reason, but I propose to leave it like it is. I don't think it's too confusing and moving them all down is quite time consuming :) Please let me know if it was confusing for you!

    .. code-block:: yaml

        # translations/messages+intl-icu.en.yaml
        say_hello: 'Hello {name}!'

    .. code-block:: xml

>         <!-- translations/messages+intl-icu.en.xlf -->

Reviewer 2: In `trabslation/debug.rst` we are using `.xliff` as extension:
```rst
.. configuration-block::

    .. code-block:: xml

        <!-- translations/messages.fr.xliff -->
        <?xml version="1.0"?>
        <xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2">
            <file source-language="en" datatype="plaintext" original="file.ext">
                <body>
                    <trans-unit id="1">
                        <source>Symfony is great</source>
                        <target>J'aime Symfony</target>
                    </trans-unit>
                </body>
            </file>
        </xliff>
```

But in general we a re using `.xlf`, which one should we communicate in the docs?

Reviewer: I think `xlf` is the recommended one. We're using it in the Symfony Demo too: https://github.com/symfony/demo/tree/master/translations

Reviewer 2: I will change the other occurrences then in a separate PR
